### PR TITLE
Multimedia as aggregate root entity

### PR DESCRIPTION
**Pull request by Daniel-Marynicz**

### CODE REVIEW

File: features/multimedia.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Feature: Multimedia
+  In order to mange Multimedia
+  I need abe able to create and retrieve through the API.
+  Scenario: Upload new multimedia file
+    Given current authentication token
+    And I attach "features/multimedia-test-image.png" to the request as upload
+    When I request "/api/v1/multimedia/upload" using HTTP POST
+    Then the response code is 201
```
Comment (BastekBielawski) [inline features/multimedia.feature]: `Then created response is received` we have

File: module/multimedia/migrations/Version20180807065948.php [hunk 2]
```diff
@@ -29,9 +31,50 @@ public function up(Schema $schema): void
...
+                'MULTIMEDIA_UPDATE',
+                'MULTIMEDIA_DELETE',
+            ]
+        );
+
+        $this->createEventStoreEvents([
+            MultimediaCreatedEvent::class => 'Multimedia added',
+        ]);
+    }
+
```
Comment (BastekBielawski) [inline module/multimedia/migrations/Version20180807065948.php]: 1 empty line, not 2

File: module/multimedia/src/Domain/Entity/Multimedia.php [hunk 3]
```diff
@@ -131,4 +131,14 @@ public function getCrc(): string
     {
         return $this->crc;
     }
+
+    protected function applyMultimediaCreatedEvent(MultimediaCreatedEvent $event)
```
Comment (BastekBielawski) [inline module/multimedia/src/Domain/Entity/Multimedia.php]: PHPDoc

File: module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        int $size,
+        string $crc,
+        ?string $mime = null
+    ) {
+        $this->id = $id;
+        $this->name = $name;
+        $this->extension = $extension;
+        $this->mime = $mime;
+        $this->size = $size;
+        $this->crc = $crc;
```
Comment (BastekBielawski) [inline module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php]: What kind of CRC? Maybe VO will be good solution?

File: module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        string $extension,
+        int $size,
+        string $crc,
+        ?string $mime = null
+    ) {
+        $this->id = $id;
+        $this->name = $name;
+        $this->extension = $extension;
+        $this->mime = $mime;
+        $this->size = $size;
```
Comment (BastekBielawski) [inline module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php]: Size in bytes, kB, MB? @rprzedzik what do you think about VO for this?

File: module/multimedia/src/Persistence/Dbal/Projector/MultimediaCreatedEventProjector.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * @param Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @param DomainEventInterface $event
```
Comment (BastekBielawski) [inline module/multimedia/src/Persistence/Dbal/Projector/MultimediaCreatedEventProjector.php]: `@inheritdoc`

File: module/multimedia/src/Persistence/Dbal/Projector/MultimediaCreatedEventProjector.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+     * @return bool
+     */
+    public function supports(DomainEventInterface $event): bool
+    {
+        return $event instanceof MultimediaCreatedEvent;
+    }
+
+    /**
+     * @param AbstractId           $aggregateId
```
Comment (BastekBielawski) [inline module/multimedia/src/Persistence/Dbal/Projector/MultimediaCreatedEventProjector.php]: `@inheritdoc`

File: features/multimedia.feature [hunk 8]
```diff
@@ -0,0 +1,17 @@
+Feature: Multimedia
+  In order to mange Multimedia
+  I need abe able to create and retrieve through the API.
+  Scenario: Upload new multimedia file
```
Comment (BastekBielawski) [inline features/multimedia.feature]: Test scenario with:
- without auth token
- without uploaded file
- with empty file (we should valid it?)

File: features/multimedia.feature [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    Given current authentication token
+    And I attach "features/multimedia-test-image.png" to the request as upload
+    When I request "/api/v1/multimedia/upload" using HTTP POST
+    Then the response code is 201
+    And the response body contains JSON:
+    """
+      {"id": "01730e8d-fb8d-5afe-9be1-b621bacbb6dd"}
+    """
+  Scenario: Download uploaded multimedia file
```
Comment (BastekBielawski) [inline features/multimedia.feature]: Test scenario with:
- without auth token
- with invalid uuid
- with not existing uuid

File: features/bootstrap/ApiContext.php [hunk 10]
```diff
@@ -183,6 +183,15 @@ public function assertResponseGrid(): void
         $this->assertJsonObjectContainsKeys('configuration,columns,collection,info');
     }
 
+    /**
+     * @Then /^print last api response$/
+     */
+    public function printLastApiResponse(): void
```
Comment (BastekBielawski) [inline features/bootstrap/ApiContext.php]: This method will be for debug?

File: module/multimedia/src/Persistence/Dbal/Repository/Factory/MultimediaIdFactory.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+declare(strict_types = 1);
+
+namespace Ergonode\Multimedia\Persistence\Dbal\Repository\Factory;
+
+use Ergonode\Multimedia\Domain\Entity\MultimediaId;
+use Ramsey\Uuid\Uuid;
+
+/**
+ */
+class MultimediaIdFactory
```
Comment (BastekBielawski) [inline module/multimedia/src/Persistence/Dbal/Repository/Factory/MultimediaIdFactory.php]: @rprzedzik What do you think about that?

File: features/bootstrap/ApiContext.php (hunk 10 above)
Comment (Daniel-Marynicz) [reply features/bootstrap/ApiContext.php]: Yes it is for debug.

Comment (Daniel-Marynicz) [reply features/bootstrap/ApiContext.php]: Ok we can create ApiDebugContext, but if we have to move more than this one method, then we can crate a new task for this.

File: features/multimedia.feature (hunk 8 above)
Comment (Daniel-Marynicz) [reply features/multimedia.feature]: done

File: module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php (hunk 4 above)
Comment (Daniel-Marynicz) [reply module/multimedia/src/Domain/Event/MultimediaCreatedEvent.php]: done, phpdoc

**Pull request merged**